src/main/java/bio/terra/common/FutureUtils.java

```diff
@@ -1,87 +1,90 @@
 package bio.terra.common;
 
-import bio.terra.app.controller.exception.ApiException;
+import bio.terra.common.exception.ApiException;
+import bio.terra.common.exception.ErrorReportException;
+import bio.terra.common.exception.ServiceUnavailableException;
+import com.google.common.annotations.VisibleForTesting;
 import java.time.Duration;
 import java.util.List;
-import java.util.Optional;
+import java.util.Objects;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.Future;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.TimeoutException;
 import java.util.concurrent.atomic.AtomicReference;
-import java.util.stream.Collectors;
-import java.util.stream.Stream;
+import javax.validation.constraints.NotNull;
 
 public final class FutureUtils {
 
   private FutureUtils() {}
 
-  public static <T> List<T> waitFor(final List<Future<T>> futures) {
-    return waitFor(futures, Optional.empty());
+  /**
+   * Wait indefinitely for a list of {@link Future} objects to complete and returns a list of the
+   * resolved values. If any of the futures fail, throws the first found instance of failure
+   * (non-deterministic) and cancel any active futures remaining.
+   *
+   * @param futures The {@link List} of futures to wait for
+   * @param <T> The type that should be returned for the list of futures
+   * @return The resolved values of the futures with nulls removed. There is no guarantee that the
+   *     order of returned values matches the order of the passed in futures
+   */
+  public static <T> List<@NotNull T> waitFor(final List<Future<T>> futures) {
+    return waitFor(futures, null);
```

Comment:
Good shout -- Phil had given some earlier feedback that the usage of Optionals in this method obfuscated intent rather than clarifying it, which I agree with.